### DISCUSSION TITLE
Proposal: Reduce namespace pollution

### DISCUSSION THREAD

**Discussion by tilkinsc** [dotnet/csharplang#7255] [Language Ideas]
Reason for interest:
I love `<ImplicitUsings>enable</ImplicitUsings>`. I am a user of vscode. I use Lists and other things from the standard library in almost every file. This reduces the amount of clutter I have at the top of my file. Although, a new problem arises as a trade-off. Namespace pollution. My intellisense is overloaded with sheer amount of toys I get to play with, and I do know I am better off not having implicit usings. However, I do feel inclined to use the most basic features of the language library without having to import every last one and its subnamespaces. I am, however, a lot simpler of a creature than the implicit usings that are provided to me and many go unused.

What other languages do and comparison:
Earlier I was contrasting C# to C++ with a buddy of mine. One of the most powerful things C++ has to offer is `using namespace [namespace];` which will inject a namespace into a function. As you know, not every function needs everything in a namespace. The benefit C++ gains is having an intellisense that is geared towards what the function sees, rather than every function having the namespace. It may sound trivial, but namespaces can be quite large. A stark difference between C++ and C# is C++ allows for members of a namespace to be something other than a class, while C# only allows classes at namespace scopes.

Users of C and C++ enjoy being able to forward declare their functions and variables and types, typically wrapped up in a header for superior control of inclusion. There are more specific details of C and C++, but as an ignorant overview, the intellisense knows what it has access to via the header files located in another header, the current header, or where the files in which the definitions live (.c, .cpp). C# doesn't need to worry about forward declarations, but the intellisense needs to be able to peer into a class to see all of its members.

Proposal:
I think its time to ditch only having the file-scoped using's and allow `using [namespace];` to be inside functions to allow us fine grain control of our intellisense and what the function sees. I think its time to implement a user customizable ImplicitUsings csproj variable in which we can dictate what namespaces each file should have. Off topic and big breaking change, I think its time to reanalyze the namespace structure so that we are correctly encapsulating features of the common library.

**Comment by HaloFour** [comment]
> I think its time to implement a user customizable ImplicitUsings csproj variable in which we can dictate what namespaces each file should have.

"Implicit usings" aren't really a language feature.  They are a build system feature that causes a file to be silently emitted into the compilation unit that contains global using directives.

If you want to dictate what namespaces are imported into each file, that's what individual `using` directives are for.  You can put them into a `#region` if you want to collapse them so that you don't see them at the top of every file.

**Reply by tilkinsc** [reply]
You missed the point, the point is making namespaces global outside of code.

 Its not that I don't want to see a long list of usings, its that the ones that are in every single file aren't really constructive. Having Collections, for example, in every file just intertwines with all the other usings where it just becomes useless bulk "nonconstructive".

**Reply by CyrusNajmabadi** [reply]
@tilkinsc Please provide more information in your responses to support a constructive dialog.  There's not much a person can do with the current response as it fails to actually clarify what point it is you think was missed.
 
Thanks!

**Reply by CyrusNajmabadi** [reply]
>  its that the ones that are in every single file aren't really constructive

Sure... but it's unclear to me what solution you want here.

First, if you think those usings aren't constructive, you have two choices:

1. Do not have those usings at all.
2. Make those usings into a global using, so it can be declared once, but not specified in all files.

But it seems like you don't like '2', because it means you now get those options everywhere.  But, first, that's an IDE issue, not a language issue.  Second, i can't reconcile this with you actually liking this specified in a single place so you don't need it everywhere.

Finally, i don't see how your solutions make this better.  Specifically:

> I think its time to ditch only having the file-scoped using's and allow using [namespace]; to be inside functions to allow us fine grain control of our intellisense and what the function sees.

But this directly contradicts your point about "the ones that are in every single file aren't really constructive".  Now you'd not only go from `usings` in many files, to usings in `N` times as many methods.  This would vastly *increase* the number of usings, not decrease them.

>  I think its time to implement a user customizable ImplicitUsings csproj variable in which we can dictate what namespaces each file should have. 

This seems actively undesirable.  Now instead of just placing the usings in the file, each file has different usings, and you have to go into the project file to even know what they are on a per-file basis.  that seems *very* confusing.  At least with global/implicit-usings, it's just set once and you now have that for all the other files.  If you want per-file usings... then just put those usings in teh file.  

> Off topic and big breaking change

Please open a new discussion. :)

**Reply by tilkinsc** [reply]
I don't want usings to only be in the top. Yes, it will probably increase the number of using if you use the same logic. This could easily be resolved by allowing them at class scope as well. Intellisense goes hand in hand with the language, so I wouldn't call it an IDE issue. Yes, I do not like 2, as some things aren't fit to be used globally. They just clutter your namespace. A `using [namespace];` inside a function would let that namespaces members be visible to only that function, similiarly a class would have access to the namespaces members for all functions. It would be an optional feature of course, so if you like file scoped namespaces you wouldn't have to change anything. No, it would be bad practice to use `using [namespace];` in many methods. Its for those one time offs where you need to pull in something large but you don't want everything to use it. It's clean when used well. No, you misunderstood about a user customizable ImplicitUsings; using would be defined for all files. Having global usings isn't really clean imho.

**Comment by ufcpp** [comment]
https://github.com/dotnet/csharplang/discussions/218

**Reply by tilkinsc** [reply]
Thanks for referencing this. I guess I missed it. :)
